[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76788
* #76787
* __->__ #76518

Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

dzdang committed:
[quant][core][gpu][improvement] Enabled broadcasting multiplication
support for requantize_multiplier_tensor in quantized cudnn add, linear,
and conv2d ops

Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993580).

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication"


support for requantize_multiplier_tensor in quantized cudnn add, linear,
and conv2d ops

Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993580).

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993580).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/BinaryOps.cpp:
```diff
@@ -109,8 +109,10 @@ Tensor add(Tensor qa, Tensor qb, double output_scale, int64_t output_zero_point)
                                                             output_scale, output_zero_point, memory_format);
-  // TODO: When cudnn enables support for broadcasting, we can remove this tensor
-  at::Tensor requantize_multiplier_tensor = at::empty(quantized_output.sizes(), at::device(at::kCUDA).dtype(at::kFloat), memory_format);
+  // We will employ broadcasting scalar multiplication in cudnn in the requant_op below. For this to work, cudNN requires
```
do we want to have a utility function for this?

dzdang replied:
for creating a scalar tensor for an arbitrary number of dimensions? I don't think so

dzdang replied:
@jerryzh168 for conv, `requantize_multiplier_tensor` size is known at compile time, so I use `std::array`, but for linear and add, size is not known at compile time, so I have to use `at::SmallVector`. if I make a utility function that all 3 ops can use, I would make conv's `requantize_multiplier_tensor` use ``at::SmallVector` instead of `std::array` -- probably a small hit on performance associated with this

jerryzh168 replied:
a utility function to create this requantize_multipler_tensor

dzdang replied:
@jerryzh168 done. can you reapprove so I can land the stack?

jerryzh168 reviewed (changes requested):
maybe write a utility function that returns a requantize_multiplier_tensor?

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993580).

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993580).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/utils.h:
```diff
@@ -192,6 +192,17 @@ struct TORCH_API PackedConvWeightCudnn : public ConvPackedParamsBase<kSpatialDim
+// pointwise multiplication operations. the only reason why we need this right now is
+// we use broadcasting scalar multiplication in conv, linear, and add ops, and cuDNN requires
+// the scalar to be a scalar tensor with the same number of dimensions (num_dim) as the tensor we're multiplying to
+at::Tensor get_requant_multiplier_tensor(double requant_multiplier, uint8_t num_dim) {
```
nit: maybe align the naming, looks like we should be using getReauntMultiplerTensor from the other names in this file

jerryzh168 reviewed (approved):
lg, thanks, had a nit comment

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Enabled broadcasting multiplication support for requantize_multiplier_tensor in quantized cudnn add, linear, and conv2d ops"


Summary:
Previously, requantize_multiplier_tensor was set to be the same size as
quantized_output, as broadcasting by multiplication was not supported in
cudnn. This support has been added in cudnn as of 8.3.3. The requirement
is that requantize_multiplier_tensor still has to be a tensor, but it
can be a scalar tensor with the same number of dimensions as the tensor
it is being multiplied to.

Test Plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35993580](https://our.internmc.facebook.com/intern/diff/D35993580)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35993580).
